README.md
<p align="center">
    <img src="./img/kbAPIlogo.png?raw=true" width="200"/>
</p>

# kbAPI Server


> **This information refers to the initial deployment of the kbAPI server which is NOT complete and should NOT be used in production environments.**

## What is the kbAPI Server ?

The kbAPI server is an extension of the [kb project](https://github.com/gnebbia/kb) which exposes the knowledge base as a REST API.

It does this by wrapping the core functionality of kb using the [Flask framework](https://flask.palletsprojects.com/en/1.1.x/)

## Starting the server (MacOS and Linux)

To start the kbAPI server, simply navigate to the directory containing the `server.py` module and type

`python3 server.py`

or, to run it in the background as a detached process (thus allowing you to have the server running in the background)

`python3 server.py &`

## Starting the server (Windows)

TODO

## DOCKER container

The kbAPI serer is also available as a DOCKER container. To construct the image, use:

`docker-compose -f docker-compose-server.yml build  --force-rm`

once the image is built, it can be run with:

`docker run -p 5000:5000 -ti kb_kb  sh -c "sh"`

## Security

The server is protected by a small security framework which required login details to be passed in each call (as per REST convention).
In this initial release, there is a simple user defined: `kbuser`  with a password of `kbuser`. This will be expanded in future releases.

In order to use the server, some examples would be:

```curl 
curl --location --request GET 'http://<hostname>:<port>//list' --header 'Authorization: Basic a2J1c2VyOmtidXNlcg=='
``` 

Note that the **a2J1c2VyOmtidXNlcg==** indicates the credentials for **kbuser/kbuser**

an alternative curl call would be :
```curl
curl -u kbuser:kbuser -i http://<hostname>:<port>/list
```
A Python example using  the well-known Requests library:

```python
import requests
url = "http://<hostname>:<port>//list"
payload = {}
headers = {
  'Authorization': 'Basic a2J1c2VyOmtidXNlcg=='
}
response = requests.request("GET", url, headers=headers, data = payload)
print(response.text.encode('utf8'))

```

## Production Servers

The Docker image supplied uses the default Flask WSGI server. This is not recommended for use in production environments.
Several options are out there to replace the default server, but a recommended option is 

[![alt text][1]][2]

[1]: img/gunicornlogo.png
[2]: https://gunicorn.org/

To deploy kb-API as a gunicorn-based server, simply use:

```gunicorn -w 4 server:kbapi_app --bind "localhost:5000"```

>Note that following the gunicorn documentation, you can configure the number of worker threads, bind address, memory usage etc.

## Endpoints

| Endpoint                                      | Method | Description|
|-----------------------------------------------|-------|-------------|
| `http://<hostname>/add`                       | POST | Adds a new artifact to the current knowledge base |
| `http://<hostname>/base/current`              | GET  | Returns the current knowledge base |
| `http://<hostname>/base/delete/<base>`        | POST | Deletes the named knowledge base |
| `http://<hostname>/base/list`                 | GET  | Lists the available knowledge bases |
| `http://<hostname>/base/new/<base>`           | POST | Create the named knowledge base |
| `http://<hostname>/base/rename/<base>`        | POST | Rename the named knowledge base |
| `http://<hostname>/base/switch/<base>`        | PUT  | Switches to the named knowledge base |
| `http://<hostname>/categories`                | GET  | Lists all the current castegories in the knowledge base |
| `http://<hostname>/delete/id/<id>`            | POST | Delete a specific Artifact by ID |
| `http://<hostname>/delete/ids/<id,id,id...>`  | POST | Delete specific Artifacts by ID |
| `http://<hostname>/delete/<name>`             | POST | Delete a specific Artifact by name |
| `http://<hostname>/erase/db`                  | POST | Erase just the current knowledge base database |
| `http://<hostname>/erase/all`                 | POST | Erase all of the current knowledge base as well as files |
| `http://<hostname>/export/all`                | GET  | Export ALL the data (including files) from the current knowledge base |
| `http://<hostname>/export/kb`                 | GET  | Export JUST the data from the current knowledge base |
| `http://<hostname>/grep/<regex>`              | GET  | Returns ALL of the artifacts in the current knowledge base using the regex|
| `http://<hostname>/import   `                 | POST | Remove the current knowledge base and replace with the content of the  import file |
| `http://<hostname>/list`                      | GET  | Returns ALL of the artifacts in the current  knowledge base |
| `http://<hostname>/list/category/<category>`  | GET  | Returns artifacts in the current knowledge base in the requested category |
| `http://<hostname>/list/tags/<tags>`          | GET  | Returns artifacts in the current knowledge base which have the requested tags |
| `http://<hostname>/stats`                     | GET  | Return a JSON string of information about the current knowledge base |
| `http://<hostname>/tags `                     | GET  | Return a JSON string of all the tags in the current knowledge base |
| `http://<hostname>/templates`                 | GET  | Return a list of all the templates available in the current knowledge base|
| `http://<hostname>/template/<query>`          | GET  | Returns the list of templates that comply with the query specified as a regex |
| `http://<hostname>/template/add/<template>`   | POST | Create a new template with the content specified in the file uploaded with it |
| `http://<hostname>/template/apply/<template>` | GET  | Apply the template to a set of artifacts whose criteria meet those 
| `http://<hostname>/template/delete/<template>`| POST | Delete the specified template |
| `http://<hostname>/template/get/<template>`   | GET  | Returns the named template specified in the body of the HTTP request |
| `http://<hostname>/template/new/<template>`   | POST | Create a new named template containing the default template  |
| `http://<hostname>/template/update/<template>`| PUT  | Update the specified template |
| `http://<hostname>/update/<id>`               | PUT  | Updates an artifact by ID |
| `http://<hostname>/view/<id>`                 | GET  | View an artifact by ID  |
| `http://<hostname>/view/<title>`              | GET  | View an artifact by title  |
| `http://<hostname>/view/<category>/<name>`    | GET  | View an artifact by its name  |
| `http://<hostname>/version`                   | GET  | Returns the version of the kb software in use  |

## Endpoints which do NOT exist

The following endpoints **look like they should exist but do NOT**

| Endpoint                               | Description      |
|----------------------------------------|------------------|
| `http://<hostname>/template/edit.....` | Edit an artifact |
| `http://<hostname>/edit.....`          | Edit a template  |


All of these endpoints will return a 
```json
{
  "Method Never Allowed"
}
```

with an HTTP response code of `405`

This is due to the paradigm of the kb-API being different to the CLI-based kb i.e. there is nowhere to edit an artifact or template. The "edit" operation is effected by using a combination of `get` and `update` methods.

## API Documentation (Postman)

The complete API documentation can be found [here](https://documenter.getpostman.com/view/12840256/TVRrWQnq#intro)


## Things to be aware of

Since kb uses SQLite to store artifacts, there are small limitations which should be noted:

> 1 - SQLite can support multiple users at once. It does however lock the whole database when writing, so if there are lots of concurrent writes it is not a suitable database for the application (usually the time the database is locked is a few  milliseconds - so for most uses this does not matter). But it is very well tested and very stable (and widely used) so it can be trusted.
This means that there **may** be occasions when a write operation will fail, so this should be catered for in any application using this REST API

> 2 - If the `.../import` method is used, and the knowledge base is being used as a multi-user scenario, then ALL the artifacts will be removed, and the knowledge base will be reset to whatever is in the import file for **everyone** 


CHANGELOG.md
# Changelog
:Info: This is the changelog for kb-api.
:Author: alshapton <alshapton@gmail.com>
:Copyright: © 2020, alshapton.
:License: GPLv3 (see /LICENSE or :doc:`Appendix B <LICENSE>`.)
:Date: 2020-12-28
:Version: 0.1.0

## Version History

0.1.0 
    * Initial split from kb reepository.

server.py
# -*- encoding: utf-8 -*-
# kb v0.1.0
# A knowledge base organizer
# Copyright © 2020, alshapton.
# See /LICENSE for licensing information.

"""
kbAPI server module

:Copyright: © 2020, alshapton.
:License: GPLv3 (see /LICENSE).
"""

# Import system libraries
import os
import tempfile
from pathlib import Path

# Use the flask framework, as well as the authentication framework
from flask import Flask, make_response, request
from flask_httpauth import HTTPBasicAuth
from werkzeug.routing import BaseConverter

# Import the API functions
from kbapi.api.add import add
from kbapi.api.base import base as base_list
from kbapi.api.base import get_current, switch, make_new_base,delete_a_base,rename
from kbapi.api.list import list_cats, list_all_tags
from kbapi.api.erase import erase
from kbapi.api.delete import delete, delete_list_of_items_by_ID
from kbapi.api.export import export
from kbapi.api.grep import grep
from kbapi.api.ingest import ingest
from kbapi.api.search import search
from kbapi.api.stats import stats
from kbapi.api.template import search as search_templates
from kbapi.api.template import add as add_template
from kbapi.api.template import new as new_template
from kbapi.api.template import delete as delete_template
from kbapi.api.template import apply_on_set as apply_template
from kbapi.api.template import get_template, update_template
from kbapi.api.update import update
from kbapi.api.view import view_by_id, view_by_title, view_by_name

# Import supporting functions
from kbapi.api.constants import MIME_TYPE
from kb import db
from kb import __version__

# Get the configuration for the knowledgebase
from kb.config import get_current_base, BASE, construct_config,DEFAULT_KNOWLEDGEBASE


class ListConverter(BaseConverter):
    """
    Custom class to convert a list from a RESTful URL into a Python list
    """
    def to_python(self, value):
        """
        Split on commas
        """
        return value.split(',')


# Initialisation

# Initiate the Flask app
kbapi_app = Flask(__name__)
kbapi_app.url_map.converters['list'] = ListConverter  # Add custom converter for lists

# Initiate the authentication framework
auth = HTTPBasicAuth()

# Set Flask application parameters
DEBUG = True
PORT = 5000
HOST = '0.0.0.0'

# Methods allowed:
ALLOWED_METHODS = ['add', 'base', 'delete', 'erase', 'export', 'get', 'grep', 'ingest' 'search', 'stats', 'template', 'update', 'version', 'view']

parameters = dict(id="", title="", category="", query="", tags="", author="", status="", no_color=False, verbose=False)
# query -> filter for the title field of the artifact
# category -> filter for the category field of the artifact
# tags -> filter for the tags field of the artifact
# author -> filter for the author field of the artifact
# status -> filter for the status field of the artifact
# no_color -> determines whether  a color output is needed
# verbose -> determines if a verbose output is needed


def toJson(self):
    """
    Convert an artifct object to a JSON string
    """
    record = '{"id":%i,"title":"%s", "category":"%s","path":"%s","tags":"%s","status":"%s","author":"%s","template":"%s"}' % \
        (self.id, self.title, self.category, self.path, self.tags, self.status, self.author, self.template)

    return record


def constructResponse(results):
    """
    Constructs a response from the results obtained by a core function
    """
    response = '['
    for result in results:
        response = response + toJson(result) + ','
    response = response[:-1] + ']'
    response = response.replace('"', "'")
    return response


def construct_search_response(results, error_text):
    """
    Constructs a good/badresponse from the results obtained by a search function
    """
    if len(results) == 0:
        response = make_response(({'Error': error_text}), 404)
    else:
        response = make_response(({'Knowledge': constructResponse(results)}), 200)
    response.mimetype = MIME_TYPE['json']
    return(response)


"""
Security framework
"""


@auth.get_password
def get_password(username):
    """
    Return password for username (default).
    """
    if username == 'kbuser':
        return 'kbuser'
    return None


@auth.error_handler
def unauthorized():
    """
    Unauthorised access to the API.
    """
    resp = make_response(({'Error': 'Unauthorized access'}), 401)
    resp.mimetype = MIME_TYPE['json']
    return (resp)


"""
Error Handling
"""


@kbapi_app.errorhandler(404)
@kbapi_app.errorhandler(500)
def not_found(error):
    """
    Generic Error Handlers
    """
    error_texts = {
        404: 'Not Found',
        500: 'Internal Server Error'
    }
    error_text = error_texts.get(error.code, 'Unknown Error')
    resp = make_response(({'Error': error_text}), error.code)
    resp.mimetype = MIME_TYPE['json']
    return (resp)


"""
Pre-request tooling
"""

@kbapi_app.before_request
def for_each_request():
    """ 
    Ensure the current knowledge base is referred to
    """
    global DEFAULT_CONFIG

    # Home base for the user
    BASE = Path.home()

    # Get the current kb or DEFAULT_KNOWLEDGEBASE
    DEFAULT_CONFIG = construct_config(BASE,'')
    

"""
Routing for URLs
"""


@kbapi_app.route('/edit', methods=['GET', 'PUT', 'POST'])
@kbapi_app.route('/template/edit', methods=['GET', 'PUT', 'POST'])
@auth.login_required
def method_never_implemented():
    """
    Methods from the command line NEVER to be implemented as they do not fit the paradigm of an API
    """
    response = make_response(({'Error': 'Method Never Allowed'}), 405)
    response.allow = ALLOWED_METHODS
    response.mimetype = MIME_TYPE['json']
    return(response)


@kbapi_app.route('/add', methods=['POST'])
@auth.login_required
def add_item():
    """
    Add a new artifact to the knowledge base.
    """
    parameters["title"] = request.form.get("title", "")
    parameters["category"] = request.form.get("category", "")
    parameters["author"] = request.form.get("author", "")
    parameters["status"] = request.form.get("status", "")
    parameters["tags"] = request.form.get("tags", "")
    parameters["file"] = ""
    attachment = request.files['file']
    resp = add(args=parameters, config=DEFAULT_CONFIG, file=attachment)
    return(resp)


@kbapi_app.route('/base/current', methods=['GET'])
@auth.login_required
def list_current_base():
    """
    Return the current knowledge base
    """
    results = get_current(config=DEFAULT_CONFIG)
    return (results)


@kbapi_app.route('/base/delete/<string:name>', methods=['POST'])
@auth.login_required
def delete_base(name=''):
    """
    Delete the named knowledge base
    """
    results = delete_a_base(name,config=DEFAULT_CONFIG)
    return (results)


@kbapi_app.route('/base/list', methods=['GET'])
@auth.login_required
def list_all_bases():
    """
    List all  the knowledge bases
    """
    results = base_list(config=DEFAULT_CONFIG)
    return (results)


@kbapi_app.route('/base/new/<string:name>', methods=['POST'])
@auth.login_required
def create_new_base(name=''):
    """
    Create a new knowledge base
    """
    parameters["name"] = name
    parameters["description"] = request.form.get("description", "")

    results = make_new_base(parameters,config=DEFAULT_CONFIG)
    return (results)


@kbapi_app.route('/base/rename/<string:old>', methods=['POST'])
@auth.login_required
def rename_base(old=''):
    """
    Rename a new knowledge base
    """
    parameters["old"] = old
    parameters["new"] = request.form.get("new", "")
    parameters["description"] = request.form.get("description", "")

    results = rename(parameters,config=DEFAULT_CONFIG)
    return (results)


@kbapi_app.route('/base/switch/<string:target>', methods=['PUT'])
@auth.login_required
def switch_base(target=DEFAULT_KNOWLEDGEBASE):
    """
    Switch to a knowledge base
    """
    results = switch(target,config=DEFAULT_CONFIG)
    return (results)


@kbapi_app.route('/categories', methods=['GET'])
@auth.login_required
def list_categories():
    """
    List all  the categories
    """
    results = list_cats(config=DEFAULT_CONFIG)
    return (results)


@kbapi_app.route('/delete/<int:id>', methods=['POST'])
@auth.login_required
def delete_item_by_ID(id=''):
    """
    Delete a single artifact from the kb knowledge base.
    """
    parameters["id"] = id
    results = delete(parameters, config=DEFAULT_CONFIG)
    return (results)


@kbapi_app.route('/delete/<list:ids>', methods=['POST'])
@auth.login_required
def delete_items_by_ID(ids=''):
    """
    Delete a list of artifacts from the kb knowledge base.
    """
    return(delete_list_of_items_by_ID(ids, config=DEFAULT_CONFIG))


@kbapi_app.route('/delete/name/<string:title>', methods=['POST'])
@auth.login_required
def delete_item_by_name(title=''):
    """
    Delete an artifact from the knowledgebase by name.
    """
    parameters["title"] = title
    results = delete(parameters, config=DEFAULT_CONFIG)
    return (results)


@kbapi_app.route('/erase/<string:component>', methods=['POST'])
@auth.login_required
def erase_db(component='all'):
    """
    Erase the whole knowledgebase
    """
    results = erase(component, config=DEFAULT_CONFIG)
    return(results)


@kbapi_app.route('/export/all', methods=['GET'])
@auth.login_required
def export_kb_all():
    """
    Export the whole knowledgebase to a file
    """
    with tempfile.NamedTemporaryFile(delete=True) as f:
        params = dict()
        params["file"] = f.name
        results = export(params, config=DEFAULT_CONFIG)
        return(results)


@kbapi_app.route('/export/data', methods=['GET'])
@auth.login_required
def export_kb_data():
    """
    Export just the data to a file
    """
    with tempfile.NamedTemporaryFile(delete=True) as f:
        params = dict()
        params["file"] = f.name
        params["only_data"] = "True"
        results = export(params, config=DEFAULT_CONFIG)
        return(results)


@kbapi_app.route('/grep/<string:regex>', methods=['GET'])
@auth.login_required
def grep_artifacts(regex):
    """
    Grep the whole knowledgebase
    """
    params = dict()
    params["regex"] = regex
    params["case_insensitive"] = False
    params["no_color"] = False
    params["verbose"] = True
    results = grep(params, config=DEFAULT_CONFIG)
    response = make_response(constructResponse(results), 200)
    response.mimetype = MIME_TYPE['json']
    return(response)


@kbapi_app.route('/import', methods=['POST'])
@auth.login_required
def ingest_kb():
    """
    Import a kb export file
    """
    params = dict()
    f = request.files['file']
    return (ingest(f, params, config=DEFAULT_CONFIG))


@kbapi_app.route('/list', methods=['GET'])
@auth.login_required
def get_all():
    """
    List all  the artifacts in the knowledgebase
    """
    parameters = dict()
    results = search(parameters, config=DEFAULT_CONFIG)
    print(results)
    response = construct_search_response(results, 'There are no artifacts in the knowledgebase.')
    return(response)


@kbapi_app.route('/list/<queries>', methods=['GET'])
@auth.login_required
def get_query(queries=''):
    """
    List the artifacts matching the query
    """
    parameters["query"] = queries
    results = search(parameters, config=DEFAULT_CONFIG)
    response = construct_search_response(results, 'There are no matching artifacts.')
    return(response)


@kbapi_app.route('/list/category/<category>', methods=['GET'])
@auth.login_required
def get_category(category=''):
    """
    List the artifacts in a category
    """
    parameters["category"] = category
    results = search(parameters, config=DEFAULT_CONFIG)
    response = construct_search_response(results, 'There are no artifacts with this category.')
    return(response)


@kbapi_app.route('/list/tags/<tags>', methods=['GET'])
@auth.login_required
def get_tags(tags=''):
    """
    List the artifacts with the given tags
    """
    parameters["tags"] = tags
    results = search(parameters, config=DEFAULT_CONFIG)
    response = construct_search_response(results, 'There are no artifacts with these tags.')
    return(response)


@kbapi_app.route('/stats', methods=['GET'])
@auth.login_required
def return_stats():
    """
    Returns statistics about the knowledgebase
    """
    response = stats(DEFAULT_CONFIG)
    response.mimetype = MIME_TYPE['json']
    return(response)


@kbapi_app.route('/tags', methods=['GET'])
@auth.login_required
def list_db_tags():
    """
    List all  the tags
    """
    results = list_all_tags(config=DEFAULT_CONFIG)
    return (results)


@kbapi_app.route('/templates', methods=['GET'])
@auth.login_required
def list_all_templates():
    """
    List all the available templates
    """
    params = dict()
    return(search_templates(params, DEFAULT_CONFIG))


@kbapi_app.route('/templates/<string:templates>', methods=['GET'])
@auth.login_required
def kb_query_templates(templates):
    """
    List the templates matching a query string
    """
    params = dict()
    params["query"] = templates
    return(search_templates(params, DEFAULT_CONFIG))


@kbapi_app.route('/template/apply/<string:title>', methods=['PUT'])
@auth.login_required
def kb_apply_template(title):
    """
    Apply a template to specific artifacts
    """
    params = dict()
    params["title"] = request.form.get("title", "")
    params["category"] = request.form.get("category", "")
    params["author"] = request.form.get("author", "")
    params["status"] = request.form.get("status", "")
    params["tags"] = request.form.get("tags", "")
    params["extended_match"] = request.form.get("extended_match", "")
    return (apply_template(params, DEFAULT_CONFIG))


@kbapi_app.route('/template/new/<string:template>', methods=['POST'])
def kb_new_template(template):
    """
    Create a new template with the default template content
    """
    params = dict()
    params["template"] = template
    return(new_template(params, DEFAULT_CONFIG))


@kbapi_app.route('/template/add/<string:title>', methods=['POST'])
@auth.login_required
def kb_add_template(title):
    """
    Add a new template with custom content
    """
    params = dict()
    params["title"] = title
    attachment = request.files['file']
    return(add_template(params, DEFAULT_CONFIG, attachment))


@kbapi_app.route('/template/delete/<string:template>', methods=['POST'])
def kb_delete_template(template):
    """
    Delete a named template
    """
    params = dict()
    params["title"] = template
    return(delete_template(params, DEFAULT_CONFIG))


@kbapi_app.route('/template/get/<string:title>', methods=['GET'])
@kbapi_app.route('/template/view/<string:title>', methods=['GET'])
@auth.login_required
def kb_get_template(title):
    """
    Get/View a template
    """
    return (get_template(title, DEFAULT_CONFIG))


@kbapi_app.route('/template/update/<string:title>', methods=['PUT'])
@auth.login_required
def kb_update_template(title):
    """
    Update a named template (with content)
    """
    attachment = request.files['file']
    return (update_template(title, DEFAULT_CONFIG, attachment))


@kbapi_app.route('/update/<int:id>', methods=['PUT'])
@auth.login_required
def update_artifact(id):
    """
    Update an artifact
    """
    parameters["title"] = request.form.get("title", "")
    parameters["category"] = request.form.get("category", "")
    parameters["author"] = request.form.get("author", "")
    parameters["status"] = request.form.get("status", "")
    parameters["tags"] = request.form.get("tags", "")
    parameters["file"] = ""
    parameters["id"] = id
    attachment = request.files['file']
    resp = update(args=parameters, config=DEFAULT_CONFIG, attachment=attachment)
    return(resp)


@kbapi_app.route('/version', methods=['GET'])
@auth.login_required
def return_version():
    """
    Returns current version of the kb software
    """
    response = make_response(({'Version': str(__version__)}), 200)
    response.mimetype = MIME_TYPE['json']
    return(response)


@kbapi_app.route('/view/<int:id>', methods=['GET'])
@kbapi_app.route('/get/<int:id>', methods=['GET'])
@auth.login_required
def view_artifact_by_id(id):
    """
    Get/View an artifact by ID
    """
    conn = db.create_connection(DEFAULT_CONFIG["PATH_KB_DB"])
    return (view_by_id(conn, id, DEFAULT_CONFIG))


@kbapi_app.route('/view/<string:title>', methods=['GET'])
@kbapi_app.route('/get/<string:title>', methods=['GET'])
@auth.login_required
def view_artifact_by_title(title):
    """
    Get/View an artifact by title
    """
    conn = db.create_connection(DEFAULT_CONFIG["PATH_KB_DB"])
    return (view_by_title(conn, title, DEFAULT_CONFIG))


@kbapi_app.route('/view/<string:category>/<string:title>', methods=['GET'])
@kbapi_app.route('/get/<string:category>/<string:title>', methods=['GET'])
@auth.login_required
def view_artifact_by_name(category, title):
    """
    Get/View an artifact by category/title
    """
    conn = db.create_connection(DEFAULT_CONFIG["PATH_KB_DB"])
    return (view_by_name(conn, title, category, DEFAULT_CONFIG))


# Start the server
if __name__ == '__main__':
    kbapi_app.run(debug=DEBUG, host=HOST, port=PORT)
